lib/rouge/lexers/bibtex.rb

```diff
@@ -0,0 +1,104 @@
+# -*- coding: utf-8 -*- #
+# frozen_string_literal: true
+
+# regex based on https://github.com/SaswatPadhi/prismjs-bibtex and https://github.com/alecthomas/chroma/blob/master/lexers/b/bibtex.go
+
+module Rouge
+  module Lexers
+    class BibTeX < RegexLexer
+      title 'BibTeX'
+      desc "BibTeX"
+      tag 'bibtex'
+      filenames '*.bib'
+
+      valid_punctuation = /@\!\$\&\*\+\-\.\/\:\;\<\>\?\[\\\]\^\`\|\~/
+      valid_name = /[a-z\_#{valid_punctuation}][\w#{valid_punctuation}]*/io
+
+      state :root do
+        mixin :whitespace
+        rule %r/@comment/i, Comment
+        rule %r/@preamble/i do
+          token Name::Class
+          push :closing_brace
+          push :value
+          push :opening_brace
+        end
+        rule %r/@string/i do
+          token Name::Class
+          push :closing_brace
+          push :field
+          push :opening_brace
+        end
+        rule %r/@#{valid_name}/o do
+          token Name::Class
+          push :closing_brace
+          push :command_body
+          push :opening_brace
+        end
+        rule %r/.+/, Comment
+      end
+
+      state :opening_brace do
+        mixin :whitespace
+        rule %r/[{(]/, Punctuation, :pop!
+      end
+
+      state :closing_brace do
+        mixin :whitespace
+        rule %r/[})]/, Punctuation, :pop!
+      end
+
+      state :command_body do
+        mixin :whitespace
+        rule %r/[^\s\,\}]+/ do
+          token Name::Label
+          pop!
+          push :fields
+        end
+      end
+
+      state :fields do
+        mixin :whitespace
+        rule %r/,/, Punctuation, :field
+        rule(//) { pop! }
+      end
+
+      state :field do
+        mixin :whitespace
+        rule(/#{valid_name}/o) { token Name::Attribute; push :value; push :equal_sign }
+        rule(//) { pop! }
+      end
+
+      state :equal_sign do
+        mixin :whitespace
+        rule %r/=/, Punctuation, :pop!
+      end
+
+      state :value do
+        mixin :whitespace
+        rule %r/#{valid_name}/o, Name::Variable
+        rule %r/"/, Literal::String, :quoted_string
+        rule %r/\{/, Literal::String, :braced_string
+        rule %r/[\d]+/, Literal::Number
+        rule %r/#/, Punctuation
+        rule(//) { pop! }
+      end
+
+      state :quoted_string do
+        rule %r/\{/, Literal::String, :braced_string
+        rule %r/"/, Literal::String, :pop!
+        rule %r/[^\{\"]+/, Literal::String
+      end
+
+      state :braced_string do
+        rule %r/\{/, Literal::String, :braced_string
```

Comment:
Is this supposed to be `"`? The lexer currently allows braced strings to be defined in quoted strings but not the reverse. Or is allowing braced strings inside quoted strings above the mistake?

Reply:
Would you mind expanding your question? I don't quite get it.

Reply:
@alexpovel The effect of this rule is that if you are inside a braced string and hit an opening brace, you will enter a 'second' brace string and need _two_ closing braces to exit. This seemed like a strange result to me.

I note that [the Pygments lexer](https://github.com/pygments/pygments/blob/3f403687036fce8c9f3d49a5bb2a8bbcdc41c8ba/pygments/lexers/bibtex.py#L107) does this, though, so it looks like it is indeed intentional.